client/node_inspect.go

```diff
@@ -6,11 +6,15 @@ import (
 	"io/ioutil"
 
 	"github.com/docker/docker/api/types/swarm"
+	"github.com/pkg/errors"
 	"golang.org/x/net/context"
 )
 
 // NodeInspectWithRaw returns the node information.
 func (cli *Client) NodeInspectWithRaw(ctx context.Context, nodeID string) (swarm.Node, []byte, error) {
+	if nodeID == "" {
+		return swarm.Node{}, nil, errors.New("node id is required")
+	}
```

Comment:
I'm not sure this is needed. Other object types don't have this check (e.g. https://github.com/moby/moby/blob/81bb9978ab5ac99e84a5bf62d0d469f0aec1d506/client/service_inspect.go#L15-L34)

There's already validation in place on the CLI;

```bash
$ docker node inspect
"docker node inspect" requires at least 1 argument.
```

And the API server performs additional validation;

```bash
$ curl --unix-socket /var/run/docker.sock http://localhost/nodes/ 
{"message":"page not found"}
```

Not entirely against, but I want to guard against having validation in the client where it's not needed, and try to keep as much as possible on the daemon side

Reply:
Unfortunately I think this is necessary. I added a similar check for volumes in #34770.

The problem is in building the url. If the arg is blank, the url path ends up being `/nodes`, which is a valid URL (https://docs.docker.com/engine/api/v1.30/#operation/NodeList). I'm not sure why that curl failed.

I wish there was a better way of handling this, but I haven't found anything yet.

Reply:
Difference is in the trailing `/`; without a trailing `/`, it's "list nodes", and with a trailing `/`, it's "view/inspect node"

Reply:
Ah, right! So I guess the line below this should be using `path.Join()` to construct the url

Reply:
> Difference is in the trailing /; without a trailing /, it's "list nodes", and with a trailing /, it's "view/inspect node"

@thaJeztah My client was making a request for `nodes/` with the trailing `/`, but was receiving the results from the list nodes endpoint, so maybe there is an issue with the routing.

I found a `go-swagger` issue that seems to have removed any meaning associated with a trailing slash explaining this behavior:
https://github.com/go-swagger/go-swagger/pull/906
https://github.com/go-swagger/go-swagger/issues/899#issuecomment-276244306

This could have affected your API server's routing. You might not be seeing this on your local setup because your `Makefile` is using an old version of `go-swagger`:

```
.PHONY: swagger-gen
swagger-gen:
	docker run --rm -v $(PWD):/go/src/github.com/docker/docker \
		-w /go/src/github.com/docker/docker \
		--entrypoint hack/generate-swagger-api.sh \
		-e GOPATH=/go \
		quay.io/goswagger/swagger:0.7.4
```

While the commit that made these changes was added in version 9.0 of `go-swagger`:
https://github.com/go-swagger/go-swagger/commit/5ef60f1b60c740f0b030541c7f2ef329fe2b5a06

Meanwhile, the commit being used in the `Dockerfile` to build Docker is using a commit that points to version 0.13 of `go-wrapper`:
```
# Install go-swagger for validating swagger.yaml
ENV GO_SWAGGER_COMMIT c28258affb0b6251755d92489ef685af8d4ff3eb
RUN git clone https://github.com/go-swagger/go-swagger.git /go/src/github.com/go-swagger/go-swagger \
	&& (cd /go/src/github.com/go-swagger/go-swagger && git checkout -q $GO_SWAGGER_COMMIT) \
	&& go install -v github.com/go-swagger/go-swagger/cmd/swagger
```
https://github.com/go-swagger/go-swagger/commit/c28258affb0b6251755d92489ef685af8d4ff3eb

Here are my client/server versions:

```
Client:
 Version:      17.11.0-ce
 API version:  1.30 (downgraded from 1.34)
 Go version:   go1.9.2
 Git commit:   1caf76c
 Built:        unknown-buildtime
 OS/Arch:      darwin/amd64

Server:
 Version:      17.06.2-ce
 API version:  1.30 (minimum version 1.12)
 Go version:   go1.8.4
 Git commit:   402dd4a/17.06.2-ce
 Built:        Fri Nov 10 00:51:08 2017
 OS/Arch:      linux/amd64
 Experimental: false
```